snuba/utils/streams/kafka.py

```diff
@@ -120,22 +121,85 @@ def close(self) -> None:
-            "queued.min.messages": queued_min_messages,
-            "enable.partition.eof": False,
-        }
+) -> Mapping[str, Any]:
+    return {
+        "enable.auto.commit": False,
+        "bootstrap.servers": ",".join(bootstrap_servers),
+        "group.id": group_id,
+        "default.topic.config": {"auto.offset.reset": auto_offset_reset},
+        # overridden to reduce memory usage when there's a large backlog
+        "queued.max.messages.kbytes": queued_max_messages_kbytes,
+        "queued.min.messages": queued_min_messages,
+        "enable.partition.eof": False,
+    }
+
+
+class KafkaConsumerWithCommitLog(KafkaConsumer):
+
+    # Set of logical (not literal) offsets to not publish to the commit log.
+    # https://github.com/confluentinc/confluent-kafka-python/blob/443177e1c83d9b66ce30f5eb8775e062453a738b/tests/test_enums.py#L22-L25
+    LOGICAL_OFFSETS = frozenset(
+        [OFFSET_BEGINNING, OFFSET_END, OFFSET_STORED, OFFSET_INVALID]
     )
+
+    def __init__(
+        self,
+        configuration: Mapping[str, Any],
+        producer: ConfluentProducer,
+        commit_log_topic: str,
+    ) -> None:
+        super().__init__(configuration)
+        self.__producer = producer
+        self.__commit_log_topic = commit_log_topic
+        self.__group_id = configuration["group.id"]
+
+    def poll(self, timeout: Optional[float] = None) -> Optional[KafkaMessage]:
+        self.__producer.poll(0.0)
+        return super().poll(timeout)
+
+    def __commit_message_delivery_callback(
+        self, error: Optional[KafkaError], message: ConfluentMessage
+    ) -> None:
+        if error is not None:
+            raise Exception(error.str())
+
+    def commit(self) -> Mapping[TopicPartition, int]:
+        offsets = super().commit()
+
+        if self.__commit_log_topic:
+            for stream, offset in offsets.items():
+                # XXX: This is an abstraction leak from the Kafka consumer.
+                if offset in self.LOGICAL_OFFSETS:
+                    logger.debug(
+                        "Skipped publishing logical offset (%r) to commit log for %s",
+                        offset,
+                        stream,
+                    )
+                    continue
+                elif offset < 0:
+                    logger.warning(
+                        "Found unexpected negative offset (%r) after commit for %s",
+                        offset,
+                        stream,
+                    )
+
+                self.__producer.produce(
+                    self.__commit_log_topic,
+                    key="{}:{}:{}".format(
+                        stream.topic, stream.partition, self.__group_id
+                    ).encode("utf-8"),
+                    value="{}".format(offset).encode("utf-8"),
+                    on_delivery=self.__commit_message_delivery_callback,
+                )
+
+        return offsets
+
+    def close(self) -> None:
+        self.__producer.flush()
```

Comment:
This makes `close` an operation that can block for quite a bit longer than it would previously.

I'm a little on the fence about adding a `timeout` here: the advantage is you'd at least put a bound on the blocking of the `flush` call, but the disadvantage is that you wouldn't have any insight into whether or not the producer queue was actually flushed or not, and I am pretty strongly against adding that return value to the API of the superclass.

Ultimately, I don't think this is going to ideally encapsulated here, I guess it just depends on which scenario is more important (exiting quickly, or flushing commit logs.)

Reply:
What about throwing if the queue is not empty ? In the end, if this operation does not manage to empty the queue wouldn't it be considered an error ? In practice it is not an issue for the current usage of the commit log. Sentry would be late on some events in the synchronized consumer, but those events would be processed as soon as this consumer restarts and successfully send a new offset on the log, but in other potential usages of the log this may not be ok.
Considering the topic is compacting, I guess a consumer should assume it can miss some messages, but this is a feature ofhow the topic is configured, the commit log producer is not really supposed to depend on this configuration.

Reply:
> What about throwing if the queue is not empty ? In the end, if this operation does not manage to empty the queue wouldn't it be considered an error ?

I like this idea, I'll give the abstract `close` an optional timeout (which can be ignored by `KafkaConsumer`, since close is non-blocking there) and throw a `TimeoutError` if there are messages remaining in the produce queue when the timeout is reached. This seems like a good middle ground, thanks for the suggestion.